File: rust/dsu.rs
struct Dsu{
    parent: Vec<usize>,
    sz: Vec<usize>,
}

impl Dsu{
    fn new(n: usize) -> Self{
        let mut parent = Vec::with_capacity(n);
        for i in 0..n{
            parent.push(i);
        }
        let sz = vec![1; n];
        Dsu { parent, sz }
    }

    fn root(&mut self, x: usize) -> usize{
        if self.parent[x] == x{
            x
        }
        else{
            let r = self.root(self.parent[x]);
            self.parent[x] = r;
            r
        }
    }

    fn merge(&mut self, x: usize, y: usize){
        let mut x = self.root(x);
        let mut y = self.root(y);
        if x == y{
            return;
        }
        if self.sz[x] < self.sz[y]{
            std::mem::swap(&mut x, &mut y);
        }
        self.parent[y] = x;
        self.sz[x] += self.sz[y];
    }

    fn same(&mut self, x: usize, y: usize) -> bool{
        self.root(x) == self.root(y)
    }

    fn size(&mut self, x: usize) -> usize{
        let r = self.root(x);
        self.sz[r]
    }
}

File: new_C++/dsu.cpp
class dsu{
private:
    int _n;
    vector<int> sz;
public:
    dsu(int n) : _n(n), sz(n, -1) {}

    bool merge(int x, int y){
        assert(0 <= x && x < _n);
        assert(0 <= y && y < _n);
        x = root(x);
        y = root(y);
        if (x == y) return false;
        if (-sz[x] < -sz[y]) swap(x, y);
        sz[x] += sz[y];
        sz[y] = x;
        return true;
    }

    bool same(int x, int y){
        assert(0 <= x && x < _n);
        assert(0 <= y && y < _n);
        return root(x) == root(y);
    }

    int root(int x){
        assert(0 <= x && x < _n);
        if (sz[x] < 0) return x;
        sz[x] = root(sz[x]);
        return sz[x];
    }

    int size(int x){
        assert(0 <= x && x < _n);
        return -sz[root(x)];
    }
};
